googlecloud-plugins/src/main/java/co/cask/hydrator/plugin/sink/GCSBatchSink.java

```diff
@@ -0,0 +1,134 @@
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.etl.api.batch.BatchSinkContext;
+import co.cask.hydrator.common.ReferenceBatchSink;
+import co.cask.hydrator.common.ReferencePluginConfig;
+import com.google.common.reflect.TypeToken;
+import com.google.gson.Gson;
+import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;
+
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * {@link GCSBatchSink} that stores the data to Google Cloud Storage Bucket.
+ * @param <KEY_OUT> the type of key the sink outputs
+ * @param <VAL_OUT> the type of value the sink outputs
+ */
+public abstract class GCSBatchSink<KEY_OUT, VAL_OUT> extends ReferenceBatchSink<StructuredRecord, KEY_OUT, VAL_OUT> {
+  public static final String PROJECT_ID_DES = "Google Cloud Project ID with access to configured GCS buckets.";
+  public static final String SERVICE_KEY_FILE_DES = "The JSON certificate file of the " +
+    "service account used for GCS access.";
+  private static final String FILESYSTEM_PROPERTIES_DESCRIPTION = "A JSON string representing a map of properties " +
+    "needed for the distributed file system.";
+  private static final Gson GSON = new Gson();
+
+  private static final Type MAP_STRING_STRING_TYPE = new TypeToken<Map<String, String>>() { }.getType();
+
+  private final GCSSinkConfig config;
+
+  protected GCSBatchSink(GCSSinkConfig config) {
+    super(config);
+    this.config = config;
+    // Update the fileSystemProperties to include the projectId and jsonKeyFile, so that prepareRun only sets
+    // the fileSystemProperties in the configuration, and not deal with projectId and jsonKeyFile separately.
+    // Do not create file system properties if macros were provided unless in a test case.
+    if (!this.config.containsMacro("fileSystemProperties") && !this.config.containsMacro("jsonKeyFile") &&
+      !this.config.containsMacro("projectId")) {
+      this.config.fileSystemProperties = this.config.getFileSystemProperties(this.config.fileSystemProperties,
+                                                                             this.config.projectId,
+                                                                             this.config.jsonKeyFile);
+    }
+  }
+
+  @Override
+  public final void prepareRun(BatchSinkContext context) {
+    OutputFormatProvider outputFormatProvider = createOutputFormatProvider(context, config.fileSystemProperties);
+    context.addOutput(Output.of(config.referenceName, outputFormatProvider));
+  }
+
+  protected abstract OutputFormatProvider createOutputFormatProvider(BatchSinkContext context,
+                                                                     String fileSystemProperties);
+
+  /**
+   * GCS Sink configuration.
+   */
+  public static class GCSSinkConfig extends ReferencePluginConfig {
+
+    @Name("bucketKey")
+    @Description("GCS Bucket to use to store the data")
```

Comment:
It makes sense to have all these strings in one place rather than having some here and some above. I would replace this with a String as the others.